### DISCUSSION TITLE
What asynchronous runtime does bevy use for concurrency?

### DISCUSSION THREAD

**Discussion by Bjohnson131** [bevyengine/bevy#9287] [Q&A] [answered]
Pretty much as it says on the tin. 

I don't see the common rust asynchronous runtimes in cargo.toml (tokio, rayon, futures), yet on the front page of the bevy website, Bevy advertises "Lock-Free Parallel Scheduler", which would indicate that an asynchronous runtime is used for processing. What runtime does Bevy use for concurrency?

**Comment by killercup** [comment, marked as answer]
_tl;dr_ Bevy supports regular futures and parallel iterators. It does not depend on a "full" async I/O lib like tokio.

Not an expert on the topic, just looking at the bevy_task crate. It depends on these:

https://github.com/bevyengine/bevy/blob/0566e73af460b9d128354a20970b1e1e0cc3b719/crates/bevy_tasks/Cargo.toml#L16-L20

These are the same crates used by [smol](https://github.com/smol-rs/smol#subcrates), a small alternative to tokio. The main feature is that bevy_tasks has a thread pool to run futures/tasks. The [async compute example](https://github.com/bevyengine/bevy/blob/0566e73af460b9d128354a20970b1e1e0cc3b719/examples/async_tasks/async_compute.rs) shows one way of using it, another is shown in the [parallel query example](https://github.com/bevyengine/bevy/blob/0566e73af460b9d128354a20970b1e1e0cc3b719/examples/ecs/parallel_query.rs) (this is almost like rayon but using the bevy_task thread pool).
